### PR TITLE
feat(vulnerabilities): add feature-flagged support for OSV

### PR DESCRIPTION
**Pull request by Churro**
## Changes

Adds integration for vulnerability alerts based on data renovate finds in the [osv.dev](https://osv.dev) database. This PR continues the work @JamieMagee initiated in https://github.com/renovatebot/renovate/pull/15159 and further augments it by addressing previous TODOs, test coverage and support for a larger set of ecosystems and versioning schemes.

There are 3 building blocks realized in this PR:

- Check if packages are affected by vulnerabilities
  - Done by implementing the [evaluation algorithm](https://ossf.github.io/osv-schema/#evaluation), as included in the OSV schema
    - The intention here is also to make it easier maintainable in renovate in case of future changes
    - _Side-note_: The check for `limit` is missing on purpose, as it is only used with "versions" that are Git hashes
- If vulnerable, find a version that remediates the vulnerability
- Create a package rule that introduces version constraints and vulnerability information in PR notes
  - As OSV entries describe the severity only using the CVSS vector, [vuln-vects](https://www.npmjs.com/package/vuln-vects) is used to infer the CVSS base score. 

Inline with https://github.com/renovatebot/renovate/pull/15159, the feature is opt-in and needs to be enabled using the  `osvVulnerabilityAlerts` config flag.

## Context

- https://github.com/renovatebot/renovate/issues/6562
- https://github.com/renovatebot/renovate/pull/15159

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Test repos (split because GH enforces secondary rate limits when creating > 10 PRs):
- https://github.com/renovate-demo/6562-vulndemo/pulls
  - `enabledManagers: ['bundler', 'cargo', 'composer', 'gomod', 'gradle', 'maven']`
- https://github.com/renovate-demo/6562-vulndemo2/pulls
  - `enabledManagers: ['mix', 'npm', 'nuget', 'pip_requirements', 'poetry']`

### CODE REVIEW

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [inline lib/workers/repository/process/vulnerabilities.ts:145]: I tend to think this should be: `const versioning = dep.versioning ?? getDefaultVersioning(dep.datasource);`

Similar to here: https://github.com/renovatebot/renovate/blob/5c5cde9d780626a9bc9d4d5fd33fdd33bb0d5847/lib/workers/repository/process/lookup/index.ts#L58

Would be great to get confirmation before I set this based on a vague idea...

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [inline lib/workers/repository/process/vulnerabilities.ts:370]: I think this will only work if '>' is supported by the versioning scheme (semver for example). 

E.g., when setting a package rule like:
```
matchPackageNames: ["com.google.guava:guava"],
matchCurrentVersion: "27.1-jre"
allowedVersions: "> 29.0"
```

The logs show: 
```
Falling back to npm semver syntax for allowedVersions
"depName": "com.google.guava:guava"
```
but semver coercion / satisfy (see [here](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/process/lookup/filter.ts#L83)) would then cause the `-jre` compat suffix to be "overlooked", leading to an update PR to `guava:30.0-android`.

Hence, I think I will need to check here if `>` is supported by the scheme, e.g. using `versioningApi.isValid()`. Would be great if someone could verify whether my thoughts are correct here.

File: lib/workers/repository/process/extract-update.ts
Comment (viceice) [inline lib/workers/repository/process/extract-update.ts:170]: doesn't it makes sense to move this to the imported files

File: lib/workers/repository/process/vulnerabilities.ts
Comment (secustor) [inline lib/workers/repository/process/vulnerabilities.ts:326]:
```suggestion
    return (
      this.isPackageAffected(ecosystem, packageName, affected) &&
      (this.includedInVersions(depVersion, affected) ||
        this.includedInRanges(depVersion, affected, versioningApi))
    );
```

File: lib/workers/repository/process/vulnerabilities.ts
Comment (secustor) [reply lib/workers/repository/process/vulnerabilities.ts:370]: As first step I think this is fine.    

After implementing this we should look into encapsulating the retrieval of the vulnerabilities in a function which returns a new interface. That way we could on one hand add the package rules for updates if there are fixes for them and add warnings on the dependency dashboard even if there are non currently.

**Review by secustor — CHANGES_REQUESTED** [review]
Really like to see this implemented.

File: lib/workers/repository/process/extract-update.spec.ts
Comment (viceice) [inline lib/workers/repository/process/extract-update.spec.ts:47]: i think you need to shift this one level up

File: lib/workers/repository/process/vulnerabilities.ts
Comment (viceice) [inline lib/workers/repository/process/vulnerabilities.ts:214]: this is heavy, please create a version map to speed up sorting or preprocess to separate field before sorting

File: lib/workers/repository/process/vulnerabilities.ts
Comment (viceice) [inline lib/workers/repository/process/vulnerabilities.ts:237]: why such a hacky value access? please replace with normal property access.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (viceice) [reply lib/workers/repository/process/vulnerabilities.ts:282]: sounds good , should be done in separate PR to be merged before this

File: lib/workers/repository/process/vulnerabilities.ts
Comment (viceice) [inline lib/workers/repository/process/vulnerabilities.ts:461]: are you sure we need to sanitize here? i think this is already done before posting to platform

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:464]: I think it's needed here because `sanitizeMarkdown()` is just applied on release notes, not on everything that's posted:

https://github.com/renovatebot/renovate/blob/8438e3068584ef2b5bda7bc994d0a53b3cd548fd/lib/workers/repository/update/pr/body/changelogs.ts#L37

The only other spot where it's called besides release notes, is the description text for Github vuln alerts:
https://github.com/renovatebot/renovate/blob/8438e3068584ef2b5bda7bc994d0a53b3cd548fd/lib/workers/repository/init/vulnerability.ts#L199

File: lib/workers/repository/process/extract-update.spec.ts
Comment (viceice) [reply lib/workers/repository/process/extract-update.spec.ts:47]: wait, remove it. we've enabled it globally for jest

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:239]: My challenge here is that the property names vary and the intention is to really to sort by value, no matter how the key is named. E.g., the input can look like this:

```json
[
  { "fixed": "1.7.6" },
  { "introduced": "1.6.0" },
  { "last_affected": "1.7.4" },
  { "introduced": "0.6.0" },
  { "fixed": "0.8.5" },
  { "limit": "1.2.3" }
]
```

I can't really figure out how to do it differently....

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts]: As a preliminary step, I've now extracted the _greater than or equal to_ and just _greater than_ checks into separate helper methods.

It's not clear to me how (and where) these would fit into the versioning API:
- The checks utilize existing methods and work generic to the underlying versioning scheme. Would adding to the `versioningApi` mean that every scheme needs to implement it. 
- The _greater than_ check is essentially just some boilerplate around the existing `versioningApi.isGreaterThan` check due to unknown / untrusted input from the OSV db.

If you still think the checks should go elsewhere, please let me know where to place them.

File: lib/workers/repository/process/extract-update.ts
Comment (Churro) [reply lib/workers/repository/process/extract-update.ts:170]: Tests for `fetchVulnerabilities` need a lot of mocking that is specific to `extract` and `lookup` in `extract-update.ts`. I'd prefer to rather keep these together in `extract-update.spec.ts` and not replicate them to `vulnerabilities.spec.ts`, as they are irrelevant for testing the functionality of the `Vulnerabilities` class itself.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:363]: Great suggestion with showing vulnerabilities on the dashboard!

I now went ahead and changed the condition to not always blindly pick the first release. Note that this is only relevant for the subset of vulnerabilities which don't specify an exact fixed version but only the `last_affected` version. 

After further testing, my rationale is that it weighs higher that renovate proposes mergable and compatible PRs (`27.1-jre` -> `31.0-jre`), over PRs that are just one version above the no-longer-vulnerable one but faulty (`27.1-jre` -> `30.0-android`). In fact, upgrading from a vulnerable old version to the newest release shouldn't be perceived as a bug in renovate at all, so I think we should be good with keeping this for now.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (JamieMagee) [reply lib/workers/repository/process/vulnerabilities.ts:245]: Interestingly, I cannot find a single vulnerability in the OSV database that uses `limit`.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (JamieMagee) [inline lib/workers/repository/process/vulnerabilities.ts:454]: You might want to prefer `CVSS_V3` over `CVSS_V2` if both are available.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (JamieMagee) [inline lib/workers/repository/process/vulnerabilities.ts:427]: The `0` could be misleading

**Comment by JamieMagee** [comment]
Thanks for getting this across the line :tada: Can't wait to see this in Renovate.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:245]: Me neither. I just put it into the snippet above to cover potential keys.

Interesting note about limit from the [examples section ](https://ossf.github.io/osv-schema/#examples-2)in the OSV schema:
> "limit" events are typically not necessary for describing numbered (linear) version ranges and should not be used. They are more useful for git ranges, where it has more implications for the evaluation algorithm.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (HonkingGoose) [reply lib/workers/repository/process/vulnerabilities.ts:363]: > Great suggestion with showing vulnerabilities without fixes on the dashboard!

I like this idea as well, see issue:

- #20242

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1500]:
```suggestion
You will only get OSV-based vulnerability alerts for _direct_ dependencies.
```

Let's emphasize the key word _direct_ here, and use a simpler sentence.

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1501]:
```suggestion
Renovate only queries the OSV database for dependencies that use one of these datasources:
```

How about a sentence like this?

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1498]:
```suggestion
Set `osvVulnerabilityAlerts` to `true` to get pull requests with vulnerability fixes.
```

Do we need to mention "once (security fixes) are available"? If there's no security fix, there's nothing for Renovate to put in the PR. At least that is clear to me, but maybe not to others?

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1497]:
```suggestion
Renovate integrates with [OSV](https://osv.dev/), an open-source vulnerability database, to check if extracted dependencies have known vulnerabilities.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md
Comment (Churro) [reply docs/usage/configuration-options.md:1498]: Vulnerabilities are very often published when there's no fix available yet. "once they are available" was meant to prevent user questions on "why didn't renovate create a PR that fixes vulnerability ABC". Besides that, the idea was to hint that renovate *continuously* checks for fixes.

Comment (HonkingGoose) [reply docs/usage/configuration-options.md:1498]: Thank you for the background information!

```suggestion
Set `osvVulnerabilityAlerts` to `true` to get pull requests with vulnerability fixes (once they are available).
```

How about this then? 😉

**Review by JamieMagee — APPROVED** [review]

**Review by secustor — DISMISSED** [review]

**Review by HonkingGoose — DISMISSED** [review]
Docs are okay!

**Review by JamieMagee — DISMISSED** [review]

**Comment by JamieMagee** [comment]
I'll let @viceice have another review, if he wants, before merging.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (viceice) [inline lib/workers/repository/process/vulnerabilities.ts:245]: we then should make a record cache again, because those calls are very expensive.

**Comment by viceice** [comment]
forgot to send my review 😅

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:245]: In general, true. I ran a check over all 13937 advisories and found that on average this operation would sort only **4 values** (3.938 to be precise, a 0 version wouldn't be sorted).

Changing this would most probably require building a map based on object indices, since the keys can occur multiple times. In the sort operation itself, I'd then have no access to these indices, as `a` and `b` don't have this info.

So overall, I think that in this specific case the gain is simply not worth the effort and extra code.

Comment (viceice) [reply lib/workers/repository/process/vulnerabilities.ts:245]: ok, please add a comment to the code about that.

**Review by viceice — COMMENTED** [review]
otherwise LGTM.

File: lib/workers/repository/process/vulnerabilities.ts
Comment (Churro) [reply lib/workers/repository/process/vulnerabilities.ts:179]: Makes sense for me. I did a check on a dump of all supported vuln dbs to see how noisy this could be:
there are currently ~51.000 events, of which ~750 don't specify a `fixed` or `last_affected` version -> 1.5%

File: lib/workers/repository/process/extract-update.spec.ts
Comment (szpak) [inline lib/workers/repository/process/extract-update.spec.ts:31]: As a GitLab user, I'm really glad seeing that MR being finished. Looking at the changes, I see that "cache" keyword is only used in test. I wonder, if those calls will be automatically cached with configured Redis as the other repository call (to minimize the number of calls in subsequent executions)?

**Review by JamieMagee — APPROVED** [review]

File: lib/workers/repository/process/extract-update.spec.ts
Comment (JamieMagee) [reply lib/workers/repository/process/extract-update.spec.ts:31]: The entire OSV database is going to be bundled with Renovate, as part of https://github.com/renovatebot/osv-offline. There won't be any HTTP calls made at all.

**Review by viceice — APPROVED** [review]

File: lib/workers/repository/process/extract-update.spec.ts
Comment (szpak) [reply lib/workers/repository/process/extract-update.spec.ts:31]: Ok, I've missed it...

**Review by secustor — APPROVED** [review]

File: lib/workers/repository/process/extract-update.spec.ts
Comment (szpak) [reply lib/workers/repository/process/extract-update.spec.ts:31]: Sorry to dig out this old thread, but playing with this feature, I wonder, if (I'm not a JS/TS developer) using `import { Ecosystem, Osv, OsvOffline } from '@renovatebot/osv-offline'` will use the latest version of that "module" or to have the latest "database", I need to constantly update the Renovate image itself?

Comment (JamieMagee) [reply lib/workers/repository/process/extract-update.spec.ts:31]: The databases are stored in GitHub releases[^1] and a new database is downloaded if the current one is more than 1 day old:

https://github.com/renovatebot/osv-offline/blob/6ed04c96cf7937b889f98d1ad065dc252acef561/packages/osv-offline/src/lib/download.ts#L30-L35

No need to update the `osv-offline` package.

[^1]: https://github.com/renovatebot/osv-offline/releases

**Comment by icanhazstring** [comment]
@JamieMagee you mentioned tested it with `composer`.
The documentation however states only the following:
```
crate
go
hex
maven
npm
nuget
packagist
pypi
rubygems
```

So, is it even getting information for composer now?

**Comment by viceice** [comment]
packagist is the composer datasource 😉

**Comment by icanhazstring** [comment]
> packagist is the composer datasource 😉

Haha lol. Yes ofc. I totally overlooked that one. Nevermind. Maybe my brain was not functioning well at the time 😂🙈

**Comment by HonkingGoose** [comment]
@rarkins and @JamieMagee 

The docs for `osvVulnerabilityAlerts` has this admonition:

> **This feature is flagged as experimental**
> Experimental features might be changed or even removed at any time.
> To track this feature visit the following GitHub issue [#6562](https://github.com/renovatebot/renovate/issues/6562).

The admonition links to this issue, which we resolved. The source for the admonition:

https://github.com/renovatebot/renovate/blob/3eb96c9647ddb1f30b53c5bfce209c6db1612e5d/lib/config/options/index.ts#L1673-L1680

Should we drop these lines?

```diff
-    experimental: true,
-    experimentalIssues: [6562],
```

**Comment by JamieMagee** [comment]
@HonkingGoose I'd still consider the feature experimental for a little while, and would like to keep the feature flag. What about opening a new issue to centralise feedback collection?

**Comment by icanhazstring** [comment]
@JamieMagee If you mean "feature flag". Does it mean it has to be set to true while `vulnerabilityAlerts` have to be enabled as well?

Because I can't figure that out from the docs.

**Comment by JamieMagee** [comment]
@icanhazstring this feature only requires `osvVulnerabilityAlerts` to be set to `true`. The `vulnerabilityAlerts` option controls GitHub-only vulnerability alerts. I expect we'll merge them both under the `vulnerabilityAlerts` option in the near future.

**Comment by HonkingGoose** [comment]
> @JamieMagee If you mean "feature flag". Does it mean it has to be set to true while `vulnerabilityAlerts` have to be enabled as well?

The "feature flag" comment means this `experimental: true` flag in our code:

```typescript
 { 
   name: 'osvVulnerabilityAlerts', 
   description: 'Use vulnerability alerts from `osv.dev`.', 
   type: 'boolean', 
   default: false, 
   experimental: true, // <--- This is the "feature flag"
   experimentalIssues: [6562],
 }, 
```

We look for the `experimental: true` flag in our docs generating code to generate the warning message that the feature is "flagged as experimental". This way the code and the docs always match.

**Pull request merged**
